dht/range_streamer.cc

```diff
@@ -274,36 +273,45 @@ future<> range_streamer::stream_async() {
                             nr_ranges_streamed, nr_ranges_streamed + ranges_to_stream.size(), nr_ranges_total);
                     nr_ranges_streamed += ranges_to_stream.size();
                     if (_nr_rx_added) {
-                        sp.request_ranges(source, keyspace, ranges_to_stream);
+                        sp.request_ranges(source, keyspace, std::move(ranges_to_stream));
                     } else if (_nr_tx_added) {
-                        sp.transfer_ranges(source, keyspace, ranges_to_stream);
+                        sp.transfer_ranges(source, keyspace, std::move(ranges_to_stream));
                     }
                     sp.execute().discard_result().get();
-                    ranges_to_stream.clear();
                     // Update finished percentage
                     auto remaining = nr_ranges_to_stream();
                     float percentage = _nr_total_ranges == 0 ? 1 : (_nr_total_ranges - remaining) / (float)_nr_total_ranges;
                     _stream_manager.local().update_finished_percentage(_reason, percentage);
                     logger.info("Finished {} out of {} ranges for {}, finished percentage={}",
                             _nr_total_ranges - remaining, _nr_total_ranges, _reason, percentage);
                 };
+                dht::token_range_vector ranges_to_stream;
                 try {
+                  if (range_vec.size() <= nr_ranges_per_stream_plan) {
+                    // Do not exchange range_vec with {} for exception safety reasons.
+                    // ip_range_vec.second is used in nr_ranges_to_stream
+                    // to calculate nr_ranges_remaining, so it should be cleared
+                    // only on success.
```

Comment:
do_streaming(std::move(range_vec));
range_vec.clear();

Here, we move range_vec before it has finished successfully.  ragne_vec.size will return zero even if do_streaming has failed.

Reply:
Not any more, since do_streaming accepts and passes its `ranges_to_stream` parameter as an rvalue reference so the move takes place eventually only in `request_ranges -> add_stream_request` or `transfer_ranges -> add_transfer_ranges` and those were made exception safe in this series.

Reply:
okay, removed the 'optimize small range_vec' patch from v5